File: app/lib/features/chat/models/room.dart
import 'package:equatable/equatable.dart';

class Room extends Equatable {
  final String id;
  final String name;
  final String? type; // 'group' or 'dm'
  final String? avatarUrl;
  final String? ownerId;
  final DateTime createdAt;
  final String? lastMessage;
  final DateTime? lastMessageTime;
  final int unreadCount;
  final String? lastMessageType;
  final String? lastMessageSenderId;
  final DateTime? lastReadAt;

  const Room({
    required this.id,
    required this.name,
    this.type,
    this.avatarUrl,
    this.ownerId,
    required this.createdAt,
    this.lastMessage,
    this.lastMessageType,
    this.lastMessageSenderId,
    this.lastMessageTime,
    this.unreadCount = 0,
    this.lastReadAt,
  });

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      id: json['id'] ?? '',
      name: json['name'] ?? 'Unknown Room',
      type: json['type'],
      avatarUrl: json['avatar_url'],
      ownerId: json['owner_id'],
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      lastMessage: json['last_message'],
      lastMessageType: json['last_message_type'],
      lastMessageSenderId: json['last_message_sender_id'],
      lastMessageTime: json['last_message_time'] != null
          ? DateTime.tryParse(json['last_message_time'])
          : null,
      unreadCount: json['unread_count'] ?? 0,
      lastReadAt: json['last_read_at'] != null
          ? DateTime.tryParse(json['last_read_at'])
          : null,
    );
  }

  Room copyWith({
    String? id,
    String? name,
    String? type,
    String? avatarUrl,
    String? ownerId,
    DateTime? createdAt,
    String? lastMessage,
    String? lastMessageType,
    String? lastMessageSenderId,
    DateTime? lastMessageTime,
    int? unreadCount,
    DateTime? lastReadAt,
  }) {
    return Room(
      id: id ?? this.id,
      name: name ?? this.name,
      type: type ?? this.type,
      avatarUrl: avatarUrl ?? this.avatarUrl,
      ownerId: ownerId ?? this.ownerId,
      createdAt: createdAt ?? this.createdAt,
      lastMessage: lastMessage ?? this.lastMessage,
      lastMessageType: lastMessageType ?? this.lastMessageType,
      lastMessageSenderId: lastMessageSenderId ?? this.lastMessageSenderId,
      lastMessageTime: lastMessageTime ?? this.lastMessageTime,
      unreadCount: unreadCount ?? this.unreadCount,
      lastReadAt: lastReadAt ?? this.lastReadAt,
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    type,
    avatarUrl,
    ownerId,
    createdAt,
    lastMessage,
    lastMessageTime,
    unreadCount,
    lastReadAt,
    lastMessageSenderId,
  ];
}

class User extends Equatable {
  final String id;
  final String username;
  final String email;
  final String? phoneNumber;
  final String? firstName;
  final String? lastName;
  final String? bio;
  final String? avatarUrl;

  const User({
    required this.id,
    required this.username,
    required this.email,
    this.phoneNumber,
    this.firstName,
    this.lastName,
    this.bio,
    this.avatarUrl,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['id'] ?? '',
      username: json['username'] ?? '',
      email: json['email'] ?? '',
      phoneNumber: json['phone_number'],
      firstName: json['first_name'],
      lastName: json['last_name'],
      bio: json['bio'],
      avatarUrl: json['avatar_url'],
    );
  }

  @override
  List<Object?> get props => [
    id,
    username,
    email,
    phoneNumber,
    firstName,
    lastName,
    bio,
    avatarUrl,
  ];
}

File: backend/internal/domain/room.go
package domain

import (
	"context"
	"time"
)

// Room represents a chat room (group) or a DM conversation.
type Room struct {
	ID              string    `json:"id"`
	Name            string    `json:"name"`
	AvatarURL       string    `json:"avatar_url,omitempty"`
	OwnerID         string    `json:"owner_id,omitempty"`
	Members         []string  `json:"members,omitempty"` // List of UserIDs
	Type            string    `json:"type"`              // "group" or "dm"
	LastMessage     string    `json:"last_message,omitempty"`
	LastMessageType string    `json:"last_message_type,omitempty"` // 👉 新增這個欄位
	LastMessageTime time.Time `json:"last_message_time,omitempty"`
	UnreadCount     int       `json:"unread_count"`
	LastReadAt      time.Time `json:"last_read_at,omitempty"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

// RoomRepository defines the interface for room data persistence.
type RoomRepository interface {
	Create(ctx context.Context, room *Room) error
	GetByID(ctx context.Context, id string) (*Room, error)
	AddMember(ctx context.Context, roomID string, userID string) error
	RemoveMember(ctx context.Context, roomID string, userID string) error
	DeleteRoom(ctx context.Context, roomID string) error
	GetMembers(ctx context.Context, roomID string) ([]string, error)
	// GetUserRooms retrieves all rooms a user is a member of.
	GetUserRooms(ctx context.Context, userID string) ([]*Room, error)
	UpdateRoom(ctx context.Context, roomID string, update map[string]interface{}) error
	UpdateOwner(ctx context.Context, roomID string, newOwnerID string) error
}

// RoomUsecase defines the interface for room business logic.
type RoomUsecase interface {
	CreateRoom(ctx context.Context, name string, ownerID string, memberIDs []string) (*Room, error)
	JoinRoom(ctx context.Context, roomID string, userID string) error
	LeaveRoom(ctx context.Context, roomID string, userID string) error
	KickMember(ctx context.Context, roomID string, ownerID string, memberID string) error
	DeleteRoom(ctx context.Context, roomID string, ownerID string) error
	GetRoomMembers(ctx context.Context, roomID string) ([]string, error)
	GetUserRooms(ctx context.Context, userID string, keyword string) ([]*Room, error)
	GetRoomMedia(ctx context.Context, userID, roomID, reqType, cursor string, limit int) ([]Message, bool, error)
	UpdateRoom(ctx context.Context, roomID string, ownerID string, name *string, avatarURL *string) error
	TransferOwnership(ctx context.Context, roomID string, currentOwnerID string, newOwnerID string) error
}
